clients/vscode-hlasmplugin/src/hlasmDownloadCommands.ts

```diff
@@ -0,0 +1,697 @@
+    host: string;
+    port?: number;
+    user: string;
+    password: string;
+    secure: connectionSecurityLevel
+}): Promise<JobClient> {
+    const client = new Client();
+    client.parseList = (rawList: string): FileInfo[] => {
+        return rawList.split(/\r?\n/).slice(1).filter(x => !/^\s*$/.test(x)).map((value) => new FileInfo(value));
+    };
+    await client.access({
+        host: connection.host,
+        user: connection.user,
+        password: connection.password,
+        port: connection.port,
+        secure: connection.secure !== connectionSecurityLevel.unsecure,
+        secureOptions: connection.secure === connectionSecurityLevel.unsecure ? undefined : { rejectUnauthorized: connection.secure !== connectionSecurityLevel.rejectUnauthorized }
+    });
+
+    const checkResponse = (resp: FTPResponse) => {
+        if (resp.code < 200 || resp.code > 299)
+            throw Error("FTP Error: " + resp.message);
+    }
+    const checkedCommand = async (command: string): Promise<string> => {
+        const resp = await client.send(command);
+        checkResponse(resp);
+        return resp.message
+    }
+    const switchText = async () => { await checkedCommand("TYPE A"); }
+    const switchBinary = async () => { await checkedCommand("TYPE I"); }
+
+    await checkedCommand("SITE FILE=JES");
+    return {
+        async submitJcl(jcl: string): Promise<string> {
+            await switchText();
+            const jobUpload = await client.uploadFrom(Readable.from(jcl), "JOB");
+            checkResponse(jobUpload);
+            const jobid = /^.*as ([Jj](?:[Oo][Bb])?\d+)/.exec(jobUpload.message);
+            if (!jobid)
+                throw Error("Unable to extract the job id");
+            return jobid[1];
+        },
+        async setListMask(mask: string): Promise<void> {
+            await checkedCommand("SITE JESJOBNAME=" + mask);
+            await checkedCommand("SITE JESSTATUS=OUTPUT");
+        },
+        async list(): Promise<JobDescription[]> {
+            try {
+                await switchText();
+                return (await client.list()).map((x: FileInfo): JobDescription => {
+                    const parsedLine = /(\S+)\s+(\S+)\s+(.*)/.exec(x.name);
+                    if (!parsedLine)
+                        throw Error("Unable to parse the job list");
+                    return { jobname: parsedLine[1], id: parsedLine[2], details: parsedLine[3] }
+                });
+            }
+            catch (e) {
+                if (e instanceof FTPError && e.code == 550)
+                    return [];
+                throw e;
+            }
+        },
+        async download(target: string | Writable, id: JobId, spoolFile: number): Promise<void> {
+            await switchBinary();
+            checkResponse(await client.downloadTo(target, id + "." + spoolFile));
+        },
+        close(): void {
+            client.close();
+        }
+    };
+}
+
+export interface JobDetail {
```

Comment:
This name is confusing me. Job detail to me sounds like job attributes one can see in sysview / sdsf like class, stage, duration, cpu consumption etc. .

Not sure what would be a good name though, maybe `DownloadDetails`, `DownloadRequest`, `DownloadItem`.